File: 1370-increasing-decreasing-string/1370-increasing-decreasing-string.go
func sortString(s string) string {
	hashMap := make(map[int]int)
	for _, v := range s {
		ascii := charToAscii(v)
		hashMap[ascii]++
	}

	var answer string
	for {
		if len(hashMap) == 0 {
			break
		}

		ascendingKeys := hashMapAscendingKeys(hashMap)
		for _, v := range ascendingKeys {
			answer += string(asciiToChar(v))
			hashMap[v]--
			if hashMap[v] == 0 {
				delete(hashMap, v)
			}
		}

		descendingKeys := hashMapDescendingKeys(hashMap)
		for _, v := range descendingKeys {
			answer += string(asciiToChar(v))
			hashMap[v]--
			if hashMap[v] == 0 {
				delete(hashMap, v)
			}
		}
	}

	return answer
}

func charToAscii(char rune) int {
	return int(char - 'a')
}

func asciiToChar(ascii int) rune {
	return rune(ascii + 'a')
}

func hashMapAscendingKeys(hashMap map[int]int) []int {
	answer := make([]int, 0)
	for k := range hashMap {
		answer = append(answer, k)
	}
	sort.Slice(answer, func(i, j int) bool {
		return answer[i] < answer[j]
	})
	return answer
}

func hashMapDescendingKeys(hashMap map[int]int) []int {
	answer := make([]int, 0)
	for k := range hashMap {
		answer = append(answer, k)
	}
	sort.Slice(answer, func(i, j int) bool {
		return answer[i] > answer[j]
	})
	return answer
}

File: 1370-increasing-decreasing-string/1370-increasing-decreasing-string.java
import java.util.*;

class Solution {
    public String sortString(String s) {
        HashMap<Integer,Integer> hashMap = new HashMap<>();
        for (char ch : s.toCharArray()) {
            int acsiiValue = charToAscii(ch);
            Integer count = hashMap.getOrDefault(acsiiValue, 0);
            hashMap.put(acsiiValue, count + 1);
        }

        String answer = "";
        while (true) {
            if (hashMap.size() == 0) {
                break;
            }

            List<Integer> ascendingKeys = hashMapKeys(hashMap, true);
            for (Integer key : ascendingKeys) {
                answer += asciiToChar(key);
                Integer count = hashMap.get(key);
                if (count - 1 == 0) {
                    hashMap.remove(key);
                } else {
                    hashMap.put(key, count - 1);
                }
            }

            List<Integer> descendingKeys = hashMapKeys(hashMap, false);
            for (Integer key : descendingKeys) {
                answer += asciiToChar(key);
                Integer count = hashMap.get(key);
                if (count - 1 == 0) {
                    hashMap.remove(key);
                } else {
                    hashMap.put(key, count - 1);
                }
            }

        }
        return answer;
    }

    static public int charToAscii(char ch) {
        return ch - 'a';
    }

    static public char asciiToChar(int ascii) {
        return (char) (ascii + 'a');
    }

    static public List<Integer> hashMapKeys(HashMap<Integer,Integer> hashMap, boolean ascending) {
        List<Integer> answer = new ArrayList<>();
        for (HashMap.Entry<Integer,Integer> entry: hashMap.entrySet()) {
            answer.add(entry.getKey());
        }
        if (ascending) {
            Collections.sort(answer);
        } else {
            Collections.sort(answer, Collections.reverseOrder());
        }
        return answer;
    }
}